sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala

```diff
@@ -159,16 +161,28 @@ class JsonInferSchema(options: JSONOptions) extends Serializable with Logging {
           val bigDecimal = decimalParser(field)
             DecimalType(bigDecimal.precision, bigDecimal.scale)
         }
-        val timestampType = SQLConf.get.timestampType
         if (options.prefersDecimal && decimalTry.isDefined) {
           decimalTry.get
-        } else if (options.inferTimestamp && (SQLConf.get.legacyTimeParserPolicy ==
-          LegacyBehaviorPolicy.LEGACY || timestampType == TimestampNTZType) &&
+        } else if (options.inferTimestamp) {
+          // For text-based format, it's ambiguous to infer a timestamp string without timezone, as
+          // it can be both TIMESTAMP LTZ and NTZ. To avoid behavior changes with the new support
+          // of NTZ, here we only try to infer NTZ if the config is set to use NTZ by default.
+          if (isDefaultNTZ &&
             timestampNTZFormatter.parseWithoutTimeZoneOptional(field, false).isDefined) {
-          timestampType
-        } else if (options.inferTimestamp &&
-            timestampFormatter.parseOptional(field).isDefined) {
-          TimestampType
+            TimestampNTZType
+          } else if (timestampFormatter.parseOptional(field).isDefined) {
+            TimestampType
+          } else {
+            val utf8Value = UTF8String.fromString(field)
+            // There was a mistake that we use TIMESTAMP NTZ parser to infer LTZ type.
+            // The mistake makes it easier to infer TIMESTAMP LTZ type and we have to keep this
+            // behavior now. See SPARK-46769 for more details.
+            if (SparkDateTimeUtils.stringToTimestampWithoutTimeZone(utf8Value, false).isDefined) {
```

Comment:
In JSON, schema inference is not incremental but always start from beginning. That said, the first value doesn't matter. To avoid breaking change, I choose to keep the existing behavior in a less hacky way. This if condition is how the default NTZ parser works, which is more stable as previously the LTZ inference behavior is controlled by the option `timestampNTZFormat`

Reply:
@gengliangwang

Reply:
Shall we combine the two conditions for TimestampType?
```
else if (timestampFormatter.parseOptional(field).isDefined ||  timestampNTZFormatter.parseOptional(field).isDefined)
```
Also, it is a bit weird to use SparkDateTimeUtils.stringToTimestampWithoutTimeZone here since the previous code was using the formatters.

Reply:
It was a mistake previously, and I don't think we should keep doing it. e.g. the `timestampNTZFormatter` can have very different behavior with a user-specified pattern string. Then we infer LTZ and hit runtime error at read time because we can't really parse the value to LTZ.

Reply:
`SparkDateTimeUtils.stringToTimestampWithoutTimeZone` is safer here because it's what the default NTZ parser does.